pkg/controller/logstash/driver.go

```diff
@@ -0,0 +1,87 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package logstash
+
+import (
+	"context"
+
+	"hash/fnv"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+
+	logstashv1alpha1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/logstash/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/operator"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/tracing"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/utils/log"
+)
+
+// Params are a set of parameters used during internal reconciliation of Logstash.
+type Params struct {
+	Context context.Context
+
+	Client        k8s.Client
+	EventRecorder record.EventRecorder
+	Watches       watches.DynamicWatches
+
+	Logstash logstashv1alpha1.Logstash
+	Status   logstashv1alpha1.LogstashStatus
+
+	OperatorParams operator.Parameters
+}
+
+// K8sClient returns the Kubernetes client.
+func (p Params) K8sClient() k8s.Client {
+	return p.Client
+}
+
+// Recorder returns the Kubernetes event recorder.
+func (p Params) Recorder() record.EventRecorder {
+	return p.EventRecorder
+}
+
+// DynamicWatches returns the set of stateful dynamic watches used during reconciliation.
+func (p Params) DynamicWatches() watches.DynamicWatches {
+	return p.Watches
+}
+
+// GetPodTemplate returns the configured pod template for the associated Elastic Logstash.
+func (p *Params) GetPodTemplate() corev1.PodTemplateSpec {
+	return p.Logstash.Spec.PodTemplate
+}
+
+// Logger returns the configured logger for use during reconciliation.
+func (p *Params) Logger() logr.Logger {
+	return log.FromContext(p.Context)
+}
+
+func newStatus(logstash logstashv1alpha1.Logstash) logstashv1alpha1.LogstashStatus {
+	status := logstash.Status
+	status.ObservedGeneration = logstash.Generation
+	return status
+}
+
+func internalReconcile(params Params) (*reconciler.Results, logstashv1alpha1.LogstashStatus) {
+	defer tracing.Span(&params.Context)()
+	results := reconciler.NewResult(params.Context)
+
+	_, err := reconcileServices(params)
```

Comment:
You are ignoring the return value here. Can it simply be removed from the function?

Reply:
@kaisecheng or it could be removed and added in the upcoming PR that adds TLS support?

Reply:
We will add TLS support in the near future, so prefer to keep it. I can add a TODO to explain